compiler+runtime/include/cpp/jank/read/parse.hpp

```diff
@@ -12,33 +14,74 @@ namespace jank::runtime
 /* TODO: Rename file to processor. */
 namespace jank::read::parse
 {
-  static option<char> get_char_from_literal(native_persistent_string const &sv)
+  static result<option<native_persistent_string>, native_persistent_string>
```

Comment:
Both of these fns should be defined in the `.cpp` file, not the `.hpp` file. They're not templates. By having them be `static`, we're actually putting a separate copy of this fn into every single translation unit. We'll have over 80 copies of this fn in the jank program this way; they'll all be identical.

Reply:
Based on your questions, this needs to be explained in a few parts.

## Translation units
A translation unit (TU), in C++, is everything which compiles down to an object file (`.o`). Generally, this ties 1:1 with `.cpp` files. Each TU is compiled separately, based on the code within the `.cpp` file and all the headers included. If multiple TUs include the same header, they do all the same duplicate work to parse it.

## Linkage
Within a TU, symbols can have different types of linkage. This is what makes the symbols within the TU available to other TUs when they're linked together into a program. By default, symbols have external linkage, meaning the symbols can be accessed by other TUs when everything is linked. Furthermore, with external linkage, there may be only one definition per symbol. This is the one definition rule (ODR).

However, putting `static` on a free fn or var gives it internal linkage, which allows multiple definitions with the same name, since they don't compete in the global symbol space. Many TUs may a definition of a symbol, even differing definitions, as long as that symbol has internal linkage.

## Inline
There is another way to get around the ODR, for functions, which is to use `inline`. A function with external linkage can be marked `inline` and it will still be allowed to have multiple definitions, though still only one per TU. This allows multiple TUs to include a header which defines an inline fn without linker errors.

Function templates are implicitly inline and are generally defined in headers to be included into many TUs.

## Static
Note that the `static` keyword, in C++, has a few different meanings. Only one of them implies internal linkage. All of the meanings are noted here: https://en.cppreference.com/w/cpp/keyword/static

## Wrap
So, by having your function defined in a header and marked `static`, every TU which includes it will get its own copy, with internal linkage. This is a recipe for code bloat in the resulting binary. We have two options:

1. Mark the function `inline`, rather than `static`
2. Define the fn in a single TU and make sure it has external linkage (no `static`) and then just forward declare it in a header which is included by other TUs

Which one to choose? Well, we don't want the first one, since defining too many inline fns in headers slows down compilation. Remember, every TU which includes that header needs to process it. This takes time. In this case, it's in a common header, too. So our choice is number 2.

How does that work? Well, every TU which wants to call this fn, but which doesn't define it, will reference the symbol, but it will remain unresolved in that TU's generated `.o` file. This is ok. Once all of the `.o` files are linked together into our `jank` binary, all symbols are resolved. As long as we link in the TU which has that fn, and the fn has external linkage, we're golden. If not, we'll get linker errors.

You'll see this is the pattern that we use already in jank. For example:

* Declaration: https://github.com/jank-lang/jank/blob/main/compiler%2Bruntime/include/cpp/jank/runtime/core/seq.hpp#L172
* Definition: https://github.com/jank-lang/jank/blob/main/compiler%2Bruntime/src/cpp/jank/runtime/core/seq.cpp#L14